### PR TITLE
Uniform tracing

### PR DESCRIPTION
**Pull request by daumayr**
This PR adds uniform record and replay for all concurrency models featured in SOMns.

### Threads & Locks
Record and Replay for Locking works as usual, see CSP.
For signal and await operations the lock has to be held before the operation can be performed -> already synchronised.

The tricky part is awaitTimeout and lock reacquisitions after await returns.
awaitTimeout is handled by recording whether a given await timed out, if it timed out, we don't perform it in replay. if awaitTimeout returned before the timeout, we use a regular await to avoid nasty timing issues.

For the lock reacquisitions the easiest solution to implement (also avoids duplicating and messing with a lot of low-level implementation stuff of reentrant lock) is to use a special internal replay condition for each lock. when await returns, we check if the locks version number is right, if not we await on the internal condition. This means the lock is released again so that other threads can perform operations that have to happen before the awaiting thread continues.
The internal condition is signalled when the version number is changed, delayed wakeup from await then checks again, and may perform another await if the version is still not right.

### CEL Actors
Reimplemented the record and replay for actors with the new approach.

### STM
STM replay is based on recording the order of successful commits. global commit lock is reused for synchronization, version number only incremented on successful commits.

### CODE REVIEW

File: tests/replay/test.sh [hunk 1]
```diff
@@ -30,15 +36,35 @@ else
...
+  )
+
+  declare -a Threads=(
+    "MutexSuite.ProducerConsumer 1 0 10 4"
+    "MutexSuite.Philosophers 1 0 500 4"
+    "Lee 1 0 4 4"
+    "Vacation 1 0 7 4"
   )
+
+  declare -a STM=(
```
Comment (daumayr) [inline tests/replay/test.sh:71]: make this singlethreaded, we know there is an issue

File: src/som/interpreter/actors/Actor.java [hunk 2]
```diff
@@ -340,7 +350,8 @@ protected void execute(final ForkJoinPool actorPool) {
   }
 
   @Override
-  public void setStepToNextTurn(final boolean val) {}
+  public void setStepToNextTurn(final boolean val) {
+  }
```
Comment (daumayr) [inline src/som/interpreter/actors/Actor.java]: eclipseformat issue

File: src/som/interpreter/transactions/Transactions.java [hunk 3]
```diff
@@ -39,8 +44,10 @@
   private IdentityHashMap<SMutableArray, ArrayChange>   arrays;
 
   private static Object globalCommitLock = new Object();
+  private static long   version;
 
-  private Transactions() {}
+  private Transactions() {
+  }
```
Comment (daumayr) [inline src/som/interpreter/transactions/Transactions.java]: eclipse format error

File: src/som/primitives/processes/ChannelPrimitives.java [hunk 4]
```diff
@@ -112,7 +111,8 @@ public ActivityType getType() {
       return ActivityType.PROCESS;
     }
 
-    protected void beforeExec(final SInvokable disp) {}
+    protected void beforeExec(final SInvokable disp) {
+    }
```
Comment (daumayr) [inline src/som/primitives/processes/ChannelPrimitives.java]: eclipse format issue

File: src/som/primitives/ActivitySpawn.java [hunk 5]
```diff
@@ -202,7 +218,8 @@ protected boolean hasTagIgnoringEagerness(final Class<? extends Tag> tag) {
 
     @Child protected ExceptionSignalingNode notAValue;
 
-    @Child RecordOneEvent traceProcCreation = new RecordOneEvent(TraceRecord.PROCESS_CREATION);
+    @Child RecordOneEvent traceProcCreation =
```
Comment (daumayr) [inline src/som/primitives/ActivitySpawn.java]: make conditional

File: src/som/primitives/ActivitySpawn.java [hunk 6]
```diff
@@ -117,7 +133,7 @@ public static IsValue createIsValue() {
     @Child protected AbstractBreakpointNode onExec;
     @Child protected ExceptionSignalingNode notAValue;
     @Child RecordOneEvent                   traceProcCreation =
-        new RecordOneEvent(TraceRecord.PROCESS_CREATION);
+        new RecordOneEvent(TraceRecord.ACTIVITY_CREATION);
```
Comment (daumayr) [inline src/som/primitives/ActivitySpawn.java]: make conditional

File: src/som/primitives/threading/ConditionPrimitives.java [hunk 7]
```diff
@@ -39,12 +45,24 @@ public final Condition doCondition(final Condition cond) {
   @GenerateNodeFactory
   @Primitive(primitive = "threadingAwait:")
   public abstract static class AwaitPrim extends UnaryExpressionNode {
+
+    @Child protected static RecordOneEvent traceWakeup =
```
Comment (daumayr) [inline src/som/primitives/threading/ConditionPrimitives.java]: make conditional

File: src/som/primitives/threading/ConditionPrimitives.java [hunk 8]
```diff
@@ -39,12 +45,24 @@ public final Condition doCondition(final Condition cond) {
...
       try {
         ObjectTransitionSafepoint.INSTANCE.unregister();
         cond.await();
+        if (VmSettings.ACTOR_TRACING) {
+          TracingCondition tc = (TracingCondition) cond;
+          traceWakeup.record(tc.owner.getNextEventNumber());
+          tc.owner.replayIncrementEventNo();
+        }
+
+        // Output.println("Activity " + aid + " woke up");
```
Comment (daumayr) [inline src/som/primitives/threading/ConditionPrimitives.java]: remove comment

File: src/som/primitives/threading/ConditionPrimitives.java [hunk 9]
```diff
@@ -57,14 +75,35 @@ public final Condition doCondition(final Condition cond) {
   @GenerateNodeFactory
   @Primitive(primitive = "threadingAwait:for:")
   public abstract static class AwaitForPrim extends BinaryExpressionNode {
+    @Child protected static RecordOneEvent traceTimeout =
+        new RecordOneEvent(TraceRecord.CONDITION_TIMEOUT);
+    @Child protected static RecordOneEvent traceWakeup  =
+        new RecordOneEvent(TraceRecord.CONDITION_WAKEUP);
```
Comment (daumayr) [inline src/som/primitives/threading/ConditionPrimitives.java]: make conditional

File: src/som/primitives/threading/MutexPrimitives.java [hunk 10]
```diff
@@ -14,23 +14,53 @@
...
+import tools.replay.actors.TracingLock;
+import tools.replay.nodes.RecordEventNodes.RecordOneEvent;
 
 
 public final class MutexPrimitives {
   @GenerateNodeFactory
   @Primitive(primitive = "threadingLock:", selector = "lock")
   public abstract static class LockPrim extends UnaryExpressionNode {
+    @Child protected static RecordOneEvent traceLock =
+        new RecordOneEvent(TraceRecord.LOCK_LOCK);
```
Comment (daumayr) [inline src/som/primitives/threading/MutexPrimitives.java]: mkae conditional

File: src/som/primitives/threading/MutexPrimitives.java [hunk 11]
```diff
@@ -14,23 +14,53 @@
...
         ObjectTransitionSafepoint.INSTANCE.unregister();
-        lock.lock();
+        if (VmSettings.REPLAY) {
+          ReplayData.replayDelayNumberedEvent((TracingLock) lock);
+        }
+
+        if (VmSettings.ACTOR_TRACING) {
+          ((TracingLock) lock).tracingLock(traceLock);
+          // Output.println(
+          // "" + aid + " acquired lock v" + (((TracingLock) lock).getNextEventNumber() - 1));
```
Comment (daumayr) [inline src/som/primitives/threading/MutexPrimitives.java]: remove comments

File: src/som/primitives/threading/MutexPrimitives.java [hunk 12]
```diff
@@ -86,16 +116,29 @@ public Object critical(final ReentrantLock lock, final SBlock block) {
   @GenerateNodeFactory
   @Primitive(primitive = "threadingIsLocked:")
   public abstract static class IsLockedPrim extends UnaryExpressionNode {
+    @Child protected RecordOneEvent traceIsLocked =
+        new RecordOneEvent(TraceRecord.LOCK_ISLOCKED);
```
Comment (daumayr) [inline src/som/primitives/threading/MutexPrimitives.java]: make conditional

File: src/som/vm/Activity.java [hunk 13]
```diff
@@ -67,7 +83,8 @@ default int getNextTraceBufferId() {
    * Set the flag that indicates a breakpoint on joining activity.
    * Does nothing for non-tracing activities, i.e., when debugging is disabled.
    */
-  default void setStepToJoin(final boolean val) {}
+  default void setStepToJoin(final boolean val) {
```
Comment (daumayr) [inline src/som/vm/Activity.java]: eclipse format issue

File: src/som/vm/ObjectSystem.java [hunk 14]
```diff
@@ -502,7 +501,8 @@ private int handlePromiseResult(final SPromise promise) {
 
       try {
         Thread.sleep(500);
-      } catch (InterruptedException e) {}
+      } catch (InterruptedException e) {
```
Comment (daumayr) [inline src/som/vm/ObjectSystem.java]: eclipse format issue

File: src/tools/concurrency/TracingActors.java [hunk 15]
```diff
@@ -170,7 +186,6 @@ public ReplayActor(final VM vm) {
       super(vm);
 
       if (VmSettings.REPLAY) {
-        expectedMessages = vm.getTraceParser().getExpectedMessages(activityId);
         replayEvents = vm.getTraceParser().getReplayEventsForEntity(activityId);
 
         synchronized (actorList) {
```
Comment (daumayr) [inline src/tools/concurrency/TracingActors.java]: check if this is still used

File: src/tools/concurrency/TracingActors.java [hunk 16]
```diff
@@ -408,30 +258,22 @@ private MessageRecord peekExpected() {
...
           progress = true;
-          removeFirstExpectedMessage(a);
         } else {
           postponedMsgs.add(firstMessage);
         }
 
         if (mailboxExtension != null) {
           for (EventualMessage msg : mailboxExtension) {
-            if (!progress && a.replayCanProcess(a.nextExpectedMessage, msg)) {
+            if (!progress && a.replayCanProcess(msg)) {
```
Comment (daumayr) [inline src/tools/concurrency/TracingActors.java]: double check the progress part of the condition

Comment (daumayr) [inline src/tools/concurrency/TracingActors.java]: postponedmessages should be a priority queue?

File: src/tools/replay/TraceParser.java [hunk 17]
```diff
@@ -231,7 +227,8 @@ protected void parseTrace(final boolean scanning, final Subtrace loc,
           first = false;
         }
       }
-    } catch (IOException e) {
+    } catch (Throwable e) {
+      e.printStackTrace();
```
Comment (daumayr) [inline src/tools/replay/TraceParser.java]: undo this change

File: src/tools/replay/TraceParser.java [hunk 18]
```diff
@@ -315,69 +303,33 @@ private boolean readTrace(final boolean first, final boolean scanning, final Byt
...
-        assert b.position() == start + RecordEventNodes.TWO_EVENT_SIZE;
-        break;
-      case CHANNEL_WRITE:
-        long channelWId = b.getLong();
-        long nwrite = b.getLong();
-        ctx.currentEntity.addReplayEvent(new ChannelWriteRecord(channelWId, nwrite));
-        assert b.position() == start + RecordEventNodes.TWO_EVENT_SIZE;
-        break;
       default:
+        Output.println("MISSING");
```
Comment (daumayr) [inline src/tools/replay/TraceParser.java]: remove

File: src/tools/replay/nodes/TraceContextNode.java [hunk 19]
```diff
@@ -15,26 +15,24 @@
...
-    writeContext(process.getId(), process.getNextTraceBufferId(),
-        TraceRecord.PROCESS_CONTEXT.value);
+    writeContext(process.getId(), process.getNextTraceBufferId());
   }
 
   @Specialization
   protected void recordGeneric(final Activity activity) {
-    writeContext(activity.getId(), activity.getNextTraceBufferId(),
-        TraceRecord.ACTOR_CONTEXT.value);
+    writeContext(activity.getId(), activity.getNextTraceBufferId());
```
Comment (daumayr) [inline src/tools/replay/nodes/TraceContextNode.java]: add specializations for the other activities

File: .gitlab-ci.yml [hunk 20]
```diff
@@ -63,6 +63,15 @@ benchmark_savina_csp_job:
     - ant compile
     - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-SavinaCSP
 
+benchmark_mutex_job:
+  stage: benchmark
+  tags: [benchmarks, infinity]
+  allow_failure: true
+  script:
+    - ant compile
+    - export EXP=`if [[ "$CI_BUILD_REF_NAME" = "master" ]]; then echo "SOMns-Mutex"; else echo "SOMns-Mutex-exp"; fi`; rebench --without-nice -c --commit-id="$CI_BUILD_REF" --environment="`hostname` `cat /etc/issue | cut -d ' ' -f 1`" --project=SOMns-Mutex --branch=master codespeed.conf $EXP
```
Comment (smarr) [inline .gitlab-ci.yml]:
```suggestion
    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-Mutex
```

File: core-lib/Benchmarks/AsyncHarness.ns [hunk 21]
```diff
@@ -120,9 +120,18 @@ class AsyncHarness usingPlatform: platform = Value (
     )
 
     printRun: runTime = (
-      | stats = system traceStatistics. |
+      | stats = system traceStatistics.
+        memstats = system memoryStatistics.
+      |
       (name + ': trace size:    ' + (stats at: 1) + 'byte') println.
```
Comment (smarr) [inline core-lib/Benchmarks/AsyncHarness.ns]: Can we also have these statistics conditionally?

File: core-lib/Benchmarks/Harness.ns [hunk 22]
```diff
@@ -145,6 +145,18 @@ class Harness usingPlatform: platform = Value (
     )
 
     print: bench run: runTime = (
+      | stats = system traceStatistics.
+        memstats = system memoryStatistics.
+      |
+      (name + ': trace size:    ' + (stats at: 1) + 'byte') println.
```
Comment (smarr) [inline core-lib/Benchmarks/Harness.ns]: Can we also have these statistics conditionally?

File: core-lib/System.ns [hunk 23]
```diff
@@ -48,6 +48,8 @@ class System usingVmMirror: vmMirror = Value (
 
   (* Actor Tracing Statistics *)
   public traceStatistics = ( ^ vmMirror traceStatistics: nil )
```
Comment (smarr) [inline core-lib/System.ns:50]: Can we have a `#traceStatistics`?

File: src/som/Launcher.java [hunk 24]
```diff
@@ -58,14 +57,6 @@ private static void finalizeExecution(final int exitCode) {
       SnapshotBackend.writeSnapshot();
     }
 
-    if (exitCode != 0) {
-      ReplayActor.printMissingMessages();
-    }
-
-    if (VmSettings.MEMORY_TRACING) {
-      TracingBackend.reportPeakMemoryUsage();
```
Comment (smarr) [inline src/som/Launcher.java:66]: This can be derived from the `memoryStatistics`, right?

File: src/som/interpreter/actors/EventualSendNode.java [hunk 25]
```diff
@@ -325,6 +339,14 @@ public final Object toNearRefWithoutResultPromise(final Object[] args) {
             current.getId());
       }
 
+      if (VmSettings.REPLAY) {
+        // TODO similar thing for any other usages of actor.send(). especially in timer prim
```
Comment (smarr) [inline src/som/interpreter/actors/EventualSendNode.java]: What's this comment about? perhaps needs to be rephrased? What is "similar thing" referring to?

File: src/som/interpreter/actors/RegisterOnPromiseNode.java [hunk 26]
```diff
@@ -25,14 +37,57 @@ public void register(final SPromise promise, final PromiseMessage msg,
...
       // we need to schedule the callback/msg directly anyway
+
       synchronized (promise) {
+        if (VmSettings.REPLAY) {
+          ReplayRecord npr = current.peekNextReplayEvent();
+          if (npr.type == TraceRecord.PROMISE_CHAINED && !promise.isResolvedUnsync()) {
+            npr = current.getNextReplayEvent();
+
+            assert msg instanceof AbstractPromiseSendMessage;
+            LinkedList<ReplayRecord> events = new LinkedList<>();
```
Comment (smarr) [inline src/som/interpreter/actors/RegisterOnPromiseNode.java:54]: Any specific reason why this is a `LinkedList`?

**Review by smarr — COMMENTED** [review]
Initial comments

File: src/som/interpreter/actors/RegisterOnPromiseNode.java (hunk 26 above)
Comment (daumayr) [reply src/som/interpreter/actors/RegisterOnPromiseNode.java:54]: Looked it up, anything that maintains order is fine.

File: src/som/interpreter/actors/EventualSendNode.java (hunk 25 above)
Comment (daumayr) [reply src/som/interpreter/actors/EventualSendNode.java]: before send is used, the send event needs to be consumed and the version number attached to the message. So if we have e.g. an http server that sends messages as callbacks, it has to do that. Exceptions to this rule are when a message already has the version number stored in it, e.g. due to some promise scenarios.

File: src/som/Launcher.java (hunk 24 above)
Comment (daumayr) [reply src/som/Launcher.java:66]: yes, there is now a primitive to get the data, and printing is handled in the harnesses

File: core-lib/System.ns (hunk 23 above)
Comment (daumayr) [reply core-lib/System.ns:50]: do you mean hasTraceStatistics prim and method so that its only printed when useful?

File: codespeed.conf [hunk 27]
```diff
@@ -713,6 +739,16 @@ experiments:
...
+    SOMns-Mutex-exp:
+        description: Run mutex benchmarks
+        executions:
+            - SOMns-interp-tn:
+                suites:
+                  - mutex
+            - SOMns-graal-tn:
+                suites:
+                  - mutex
+
```
Comment (sophie-kaleba) [inline codespeed.conf]: Aren't those same benchmarks already ran a few lines above (see SOMns-Mutex)?

Comment (daumayr) [reply codespeed.conf]: the exp stuff was for the CI benchmarking in brussels, with results reported to the speed website. I guess that was deprecated by stefan and I should change it.

File: src/som/interpreter/actors/Actor.java [hunk 28]
```diff
@@ -252,8 +262,8 @@ void doRun() {
 
       t.currentlyExecutingActor = actor;
 
-      if (VmSettings.ACTOR_TRACING) {
-        ActorExecutionTrace.recordActivityContext((TracingActor) actor, tracer);
```
Comment (sophie-kaleba) [inline src/som/interpreter/actors/Actor.java]: Was this cast to `TracingActor` unnecessary in the first place? Otherwise I didn't catch a change that would explain why it is not needed anymore

File: src/som/vm/VmSettings.java [hunk 29]
```diff
@@ -63,7 +63,7 @@
     DISABLE_TRACE_FILE = getBool("som.disableTraceFile", false) || (REPLAY && !KOMPOS_TRACING);
     TRACE_SMALL_IDS = getBool("som.smallIds", false);
 
-    ACTOR_TRACING = getBool("som.actorTracing", false);
+    UNIFORM_TRACING = getBool("som.actorTracing", false);
```
Comment (sophie-kaleba) [inline src/som/vm/VmSettings.java:74]: Shouldn't the setting's name be turned into `som.uniformTracing` as well?

File: src/som/interpreter/actors/Actor.java (hunk 28 above)
Comment (daumayr) [reply src/som/interpreter/actors/Actor.java]: The tracer nodes were changed, they are now specialising on the activity type and are no longer restricted to actors. of course this is not visible here. I probably should refactor ActorExecution trace, the name isn't accurate anymore

File: src/som/vm/VmSettings.java (hunk 29 above)
Comment (daumayr) [reply src/som/vm/VmSettings.java:74]: yes, changing the flags etc would be a good idea. so from -at to -ut

File: core-lib/System.ns (hunk 23 above)
Comment (smarr) [reply core-lib/System.ns:50]: yes a hasTraceStatistics prim would be great for consistency

File: codespeed.conf (hunk 27 above)
Comment (smarr) [reply codespeed.conf]: Yeah, I removed the `exp` duplication, with the data going to ReBenchDB, that's not needed any longer.

File: src/som/interpreter/actors/EventualSendNode.java (hunk 25 above)
Comment (smarr) [reply src/som/interpreter/actors/EventualSendNode.java]: hmm, I get your explanation, but not how it relates to the comment. What exactly needs to be implemented still?

Comment (daumayr) [reply src/som/interpreter/actors/EventualSendNode.java]: Essentially it's a note on what to do when implementing external event stuff. maybe I should remove the comment and make a note somewhere else

Comment (smarr) [reply src/som/interpreter/actors/EventualSendNode.java]: Perhaps it can be rephrased to something like:
`// Note: When implementing external events, such as ... ...`

File: src/tools/replay/TraceRecord.java [hunk 30]
```diff
@@ -4,22 +4,25 @@
...
  */
 public enum TraceRecord {
-  ACTOR_CREATION(0),
-  ACTOR_CONTEXT(1),
+  ACTIVITY_CREATION(0),
+  ACTIVITY_CONTEXT(1),
   MESSAGE(2),
   PROMISE_MESSAGE(3),
   SYSTEM_CALL(4),
-  CHANNEL_CREATION(5),
```
Comment (smarr) [inline src/tools/replay/TraceRecord.java:12]: What's happening now with `CHANNEL_CREATION`?
That's not needed anymore, because it is deterministic?
Was this an artifact from the Kompos tracing for which we want it?

File: tests/replay/test.sh [hunk 31]
```diff
@@ -30,15 +44,35 @@ else
...
-  ) 
+  )
 
   declare -a Validation=(
-    "Counting 100 0 1000"
-    "Philosophers 100 0 5:5 25"
-    "DeadLock 100 0 4:2:3"
-    "Messages 100 0 1000"
-    "Sequence 100 0 100"
+    #Currently broken due to timerthread being incompatible
```
Comment (smarr) [inline tests/replay/test.sh:50]: Can I help with this, where would I start?

**Review by smarr — COMMENTED** [review]
some notes for the moment. Working through the rest still.

**Comment by smarr** [comment]
@daumayr what's the difference between the bench.conf and the codespeed.conf? do we need both?

For the codespeed.conf, I fixed some issue already, but mutex experiments look like they are not working at the moment, and might have some copy and paste mistake or so:

```
Running Benchmarks: 0.00%
 Running Benchmarks:               Philosophers (SOMns-interp-tn, mutex, 500 4) 1 None None	mean:       55.2	time left: 00:00:13: 25.00%
 Running Benchmarks:                Philosophers (SOMns-graal-tn, mutex, 500 4) 1 None None	mean:      292.0	time left: 00:00:13: 50.00%
 Executing run:ProducerConsumer (SOMns-interp-tn, mutex, 4000 4) 1 None None
     cmd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns/som -G   core-lib/Benchmarks/Harness.ns SavinaCSP.ProducerConsumer 50 0 4000 4
     cwd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns
     Run failed. Return code: 1
     Output:
         MessageNotUnderstood(SavinaCSP>>#ProducerConsumer)
         
     Execution has failed, benchmark is aborted.
         The benchmark failed 1 times in a row.
 Running Benchmarks:          ProducerConsumer (SOMns-interp-tn, mutex, 4000 4) 1 None None	mean:        0.0	time left: 00:00:04: 75.00%
 Executing run:ProducerConsumer (SOMns-graal-tn, mutex, 4000 4) 1 None None
     cmd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns/som  core-lib/Benchmarks/Harness.ns SavinaCSP.ProducerConsumer 50 0 4000 4
     cwd: /home/gitlab-runner/builds/d258e35c/0/sm951/SOMns
     Run failed. Return code: 1
     Output:
         MessageNotUnderstood(SavinaCSP>>#ProducerConsumer)
         
     Execution has failed, benchmark is aborted.
         The benchmark failed 1 times in a row.
 Running Benchmarks:           ProducerConsumer (SOMns-graal-tn, mutex, 4000 4) 1 None None	mean:        0.0	time left: 00:00:00: 100.00%
 --------------------------------------------------------------------------------------------------
   Benchmark          Executor          Suite   Extra    Core   Size   Var   #Samples   Mean (ms)  
 --------------------------------------------------------------------------------------------------
   ProducerConsumer   SOMns-graal-tn    mutex   4000 4      1                       0   Failed     
   Philosophers       SOMns-graal-tn    mutex   500 4       1                      20   292        
   ProducerConsumer   SOMns-interp-tn   mutex   4000 4      1                       0   Failed     
   Philosophers       SOMns-interp-tn   mutex   500 4       1                      20   55         
 --------------------------------------------------------------------------------------------------
```

File: .gitlab-ci.yml [hunk 32]
```diff
@@ -63,6 +63,15 @@ benchmark_savina_csp_job:
...
 
+benchmark_mutex_job:
+  stage: benchmark
+  tags: [benchmarks, infinity]
+  allow_failure: true
+  script:
+    - ant compile
+    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-Mutex
+
+
```
Comment (smarr) [inline .gitlab-ci.yml]:
```suggestion
    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-SavinaCSP
    - rebench --without-nice --experiment="CI Benchmark Run Pipeline ID $CI_PIPELINE_ID" --branch="$CI_COMMIT_REF_NAME" -c codespeed.conf SOMns-Mutex

```

Let's avoid recompiling things. Both jobs should be short enough.

File: codespeed.conf [hunk 33]
```diff
@@ -60,13 +60,13 @@ benchmark_suites:
...
             - LeeSTM:
-                extra_args: "1 0 4"
+                extra_args: "1 0 1"
                 codespeed_name: "1st.LeeSTM"
             - Vacation:
                 extra_args: "1 0 8"
                 codespeed_name: "1st.Vacation"
             - VacationSTM:
-                extra_args: "1 0 8"
+                extra_args: "1 0 1"
```
Comment (smarr) [inline codespeed.conf]: Are these changes needed?
I would prefer to keep parameters constant, if possible.
Otherwise, we can't compare over time.

File: codespeed.conf [hunk 34]
```diff
@@ -594,6 +594,22 @@ benchmark_suites:
                 extra_args: "500 4"
                 codespeed_name: "CSP.Philosophers"
 
+    mutex:
+        description: various mutex based benchmarks
+        gauge_adapter: RebenchLog
+        command: *CSP_INTERP_CMD
```
Comment (smarr) [inline codespeed.conf]: Is this the correct command?

File: src/tools/replay/nodes/RecordEventNodes.java [hunk 35]
```diff
@@ -1,13 +1,12 @@
...
 import tools.replay.TraceRecord;
-import tools.replay.actors.ActorExecutionTrace.ActorTraceBuffer;
+import tools.replay.actors.UniformExecutionTrace.UniformTraceBuffer;
 
 
 public final class RecordEventNodes {
   public static final int ONE_EVENT_SIZE   = 1 + Long.BYTES;
-  public static final int TWO_EVENT_SIZE   = 1 + (2 * Long.BYTES);
-  public static final int THREE_EVENT_SIZE = 1 + (3 * Long.BYTES);
+  public static final int THREE_EVENT_SIZE = 1 + Long.BYTES + Long.BYTES + Long.BYTES;
```
Comment (smarr) [inline src/tools/replay/nodes/RecordEventNodes.java]: This is used in only one place in the `TraceParser`. Looks like something that is not used anymore?

File: src/tools/replay/nodes/TraceContextNode.java [hunk 36]
```diff
@@ -2,11 +2,13 @@
...
+import som.primitives.threading.TaskThreads.TracedForkJoinTask;
+import som.primitives.threading.TaskThreads.TracedThreadTask;
 import som.vm.Activity;
 import tools.concurrency.TracingActors.TracingActor;
 import tools.replay.TraceRecord;
-import tools.replay.actors.ActorExecutionTrace.ActorTraceBuffer;
+import tools.replay.actors.UniformExecutionTrace.UniformTraceBuffer;
 
 
 public abstract class TraceContextNode extends TraceNode {
```
Comment (smarr) [inline src/tools/replay/nodes/TraceContextNode.java:11]: This looks like something that could be simplified with a `ClassProfile` instead of the various specializations.
It seems we really just need the concrete type to resolve the virtual method calls, right?

File: src/som/interpreter/actors/Actor.java [hunk 37]
```diff
@@ -231,15 +239,20 @@ public void doRunWithObjectSafepoints(final ActorProcessingThread t) {
...
       }
 
       while (getCurrentMessagesOrCompleteExecution()) {
         processCurrentMessages(t, dbg);
       }
+
+      if (VmSettings.UNIFORM_TRACING || VmSettings.KOMPOS_TRACING) {
+        t.swapTracingBufferIfRequestedUnsync();
+      }
+      t.currentlyExecutingActor = null;
```
Comment (smarr) [inline src/som/interpreter/actors/Actor.java]:
```suggestion
```

I think we to make `doRunWithObjectSafepoints` `private` and the `finally clause of `doRun` already does this.
What do you think?

File: src/som/interpreter/actors/ReceivedRootNode.java [hunk 38]
```diff
@@ -53,6 +55,13 @@ protected ReceivedRootNode(final SomLanguage language,
     } else {
       serializer = null;
     }
+
+    if (VmSettings.UNIFORM_TRACING) {
+      messageTracer = new RecordOneEvent(TraceRecord.MESSAGE);
+    }
+    if (VmSettings.RECEIVER_SIDE_TRACING) {
```
Comment (smarr) [inline src/som/interpreter/actors/ReceivedRootNode.java:62]: I'll assume the flag implies UNIFORM_TRACING.

File: src/som/interpreter/actors/RegisterOnPromiseNode.java [hunk 39]
```diff
@@ -71,28 +125,55 @@ public void register(final SPromise promise, final PromiseMessage msg,
...
+        this.promiseMsgSend = new RecordOneEvent(TraceRecord.PROMISE_MESSAGE);
+      }
     }
 
     public void register(final SPromise promise, final PromiseMessage msg,
         final Actor current) {
 
+      // Output.println("try register onerror " + promise.hashCode() + " in "
+      // + TracingActivityThread.currentThread().getActivity().getId());
+
```
Comment (smarr) [inline src/som/interpreter/actors/RegisterOnPromiseNode.java]:
```suggestion
```

I assume this isn't needed any longer.

File: src/som/interpreter/actors/SPromise.java [hunk 40]
```diff
@@ -335,26 +369,327 @@ public SPromise getChainedPromiseUnsync() {
...
+      this.onErrorReplay.add(msg);
+      this.version++;
+
+      tryPerformDelayedResolution();
+    }
+
+    /**
+     * Stores Message Objects in a priority Queue, to bring the messages in the original order.
+     * Use resolveReplay method before resolving promise.
+     *
```
Comment (smarr) [inline src/som/interpreter/actors/SPromise.java]:
```suggestion
```

File: src/tools/concurrency/TracingActors.java [hunk 41]
```diff
@@ -61,6 +60,29 @@ public final int getActorId() {
...
 
+    @Override
+    @TruffleBoundary
+    public synchronized void send(final EventualMessage msg,
+        final ForkJoinPool actorPool) {
+      super.send(msg, actorPool);
+      if (VmSettings.SENDER_SIDE_TRACING) {
+        msg.getTracingNode().record(this.version);
+        this.version++;
+        // TODO maybe try to get the recording itself done outside the synchronized method
```
Comment (smarr) [inline src/tools/concurrency/TracingActors.java:71]: Is this TODO still current?

File: tests/replay/test.sh [hunk 42]
```diff
@@ -74,3 +124,59 @@ do
...
+do
+  echo "$args"
+  echo "Tracing:"
+  $SOM_DIR/som -t8 -G -JXmx1500m -at core-lib/Benchmarks/AsyncHarness.ns MultiParadigmBench 1000 0 100
+  echo ""
+  echo "Replay:"
+  $SOM_DIR/som -t8 -G -JXmx1500m -vmd -r core-lib/Benchmarks/AsyncHarness.ns MultiParadigmBench 1000 0 100
+  echo ""
+  echo "========================================================"
+  echo ""
```
Comment (smarr) [inline tests/replay/test.sh]: And we might want to reduce the code duplication by using a function.

File: core-lib/Benchmarks/AsyncHarness.ns [hunk 43]
```diff
@@ -120,9 +120,21 @@ class AsyncHarness usingPlatform: platform = Value (
     )
 
     printRun: runTime = (
-      | stats = system traceStatistics. |
-      (name + ': trace size:    ' + (stats at: 1) + 'byte') println.
-      (name + ': external data: ' + (stats at: 2) + 'byte') println.
+      | stats = system traceStatistics.
+        memstats = system memoryStatistics.
+      |
+
```
Comment (smarr) [inline core-lib/Benchmarks/AsyncHarness.ns:126]: Can we avoid printing anything for which the value is 0?
I am not sure all our tools work with this, but it would minimize the output.

**Review by smarr — COMMENTED** [review]
Ok, pfew, that took a while.

File: src/tools/replay/TraceRecord.java (hunk 30 above)
Comment (smarr) [reply src/tools/replay/TraceRecord.java:12]: Is not needed for replay anymore, because they don't need their own ids.

File: src/tools/concurrency/TracingActors.java (hunk 41 above)
Comment (daumayr) [reply src/tools/concurrency/TracingActors.java:71]: I don't think this would change much, in a single threaded execution it doesn't matter. In a multithreaded execution it might reduce contention of an actor a bit, e.g. for philosophers, but i don't think the performance would change much.

File: codespeed.conf (hunk 33 above)
Comment (daumayr) [reply codespeed.conf]: The STM version is slower, the idea was to speed things up

File: core-lib/Benchmarks/AsyncHarness.ns (hunk 43 above)
Comment (daumayr) [reply core-lib/Benchmarks/AsyncHarness.ns:126]: I implemented it for the trace stats, MemoryStats have to be enabled with a flag so that shouldn't be necessary

**Comment by smarr** [comment]
With #369, we got the benchmarks already merged.
The current state of this code is pushed here, in case things go wrong: https://github.com/MetaConc/SOMns/tree/da/Uniform_Tracing_2020-08-08_Start

I'll update and force-push the branch for this PR now rebased on top of the merged #369.

File: src/som/vmobjects/SSymbol.java [hunk 44]
```diff
@@ -41,10 +40,9 @@
   public SSymbol(final String value) {
     string = value;
     numberOfSignatureArguments = determineNumberOfSignatureArguments();
-    if (VmSettings.KOMPOS_TRACING || VmSettings.ACTOR_TRACING
-        || VmSettings.TRACK_SNAPSHOT_ENTITIES) {
+    if (VmSettings.KOMPOS_TRACING || VmSettings.TRACK_SNAPSHOT_ENTITIES) {
       symbolId = (short) idGenerator.getAndIncrement();
-      TracingBackend.logSymbol(this);
```
Comment (smarr) [inline src/som/vmobjects/SSymbol.java:47]: Hmm. Does this work for Kompos?
Does the logSymbol still need to be done for it?

Comment (smarr) [reply src/som/vmobjects/SSymbol.java:47]: I restored the call. Seemed still necessary.

**Comment by smarr** [comment]
@daumayr so, to get the validation tests enabled again, we need to also handle the time threads? Do I remember that correctly?

I think, I prefer to actually merge this as is. It's huge already. And then try to do that in a new branch.
What do you think?

If that's ok for you, and my last changes are also fine with you, I'll merge this.

**Comment by daumayr** [comment]
@smarr i had a look at your changes and they look ok. 
yes, to enable the validation tests we need to have a timer that is compatible with our event infrastructure, i.e. compatible with TracingActivityThread
